pallets/moonbeam-lazy-migrations/src/foreign_asset.rs

```diff
@@ -0,0 +1,221 @@
+// Copyright 2024 Moonbeam foundation
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! # Functions for handling foreign asset migrations
+
+use super::*;
+use frame_support::sp_runtime::Saturating;
+use frame_support::traits::{fungibles::metadata::Inspect, ReservableCurrency};
+use parity_scale_codec::{Decode, Encode, MaxEncodedLen};
+use sp_core::U256;
+
+#[derive(Debug, Encode, Decode, scale_info::TypeInfo, PartialEq, MaxEncodedLen)]
+pub enum ForeignAssetMigrationStatus {
+	/// No migration in progress
+	Idle,
+	/// Migrating a foreign asset in progress
+	Migrating(ForeignAssetMigreationInfo),
+}
+
+impl Default for ForeignAssetMigrationStatus {
+	fn default() -> Self {
+		ForeignAssetMigrationStatus::Idle
+	}
+}
+
+#[derive(Debug, Encode, Decode, scale_info::TypeInfo, PartialEq, MaxEncodedLen)]
+pub(super) struct ForeignAssetMigreationInfo {
+	pub(super) asset_id: u128,
+	pub(super) remaining_balances: u32,
+	pub(super) remaining_approvals: u32,
+}
+
+impl<T: Config> Pallet<T>
+where
+	<T as pallet_assets::Config>::Balance: Into<U256>,
+	<T as pallet_asset_manager::Config>::ForeignAssetType: Into<Option<Location>>,
+{
+	/// Start a foreign asset migration by freezing the asset and creating the SC with the moonbeam
+	/// foreign assets pallet.
+	pub(super) fn do_start_foreign_asset_migration(
+		origin: OriginFor<T>,
+		asset_id: u128,
+	) -> DispatchResult {
+		ForeignAssetMigrationStatusValue::<T>::try_mutate(|status| -> DispatchResult {
```

Comment:
Shouldn't we be able to migrate multiple assets simultaneously? Looking at this, it seems that we can only migrate 1 per time.

Reply:
I introduced a change that allows approving the migration of multiple assets in one call. Once approved, the assets can then be migrated one by one. This approach ensures that each asset is migrated as quickly as possible after the migration begins, minimizing the time it remains frozen during migration.